# <center> Cleaning and Examining Ebay Car Sales Data </center>

## 1. Introduction

The aim of this project is to clean the data and analyze the included used car listings.

This is the data dictionary provided with data:

- `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
- `name` - Name of the car.
- `seller` - Whether the seller is private or a dealer.
- `offerType` - The type of listing
- `price` - The price on the ad to sell the car.
- `abtest` - Whether the listing is included in an A/B test.
- `vehicleType` - The vehicle Type.
- `yearOfRegistration` - The year in which the car was first registered.
- `gearbox` - The transmission type.
- `powerPS` - The power of the car in PS.
- `model` - The car model name.
- `kilometer` - How many kilometers the car has driven.
- `monthOfRegistration` - The month in which the car was first registered.
- `fuelType` - What type of fuel the car uses.
- `brand` - The brand of the car.
- `notRepairedDamage`- If the car has a damage which is not yet repaired.
- `dateCreated` - The date on which the eBay listing was created.
- `nrOfPictures` - The number of pictures in the ad.
- `postalCode` - The postal code for the location of the vehicle.
- `lastSeenOnline` - When the crawler saw this ad last online.

## 2. Importing modules, reading the csv and first looks at the data
Reading the csv with the right encoding (default:UTF-8) or (Latin-1 and Windows-1252)

In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv("autos.csv", encoding="Latin-1")
#print(autos.head())


Autoformating of jupyter of a variable

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [4]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

## 3. Cleaning the headers of the columns
Cleanup some of the column headers. Snakecase instead of camel case and make it easier to understand.

In [5]:
new_column_names = autos
#print("Before column header change:")
#print(new_column_names.columns)
new_column_names.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
new_column_names.columns = new_column_names.columns.str.replace("yearOfRegistration", "registration_year")
new_column_names.columns = new_column_names.columns.str.replace("monthOfRegistration", "registration_month")
new_column_names.columns = new_column_names.columns.str.replace("notRepairedDamage", "unrepaired_damage")
new_column_names.columns = new_column_names.columns.str.replace("dateCreated", "ad_created")
new_column_names.columns = new_column_names.columns.str.replace("vehicleType", "vehicle_type")
new_column_names.columns = new_column_names.columns.str.replace("offerType", "offer_type")
new_column_names.columns = new_column_names.columns.str.replace("lastSeen", "last_seen")
new_column_names.columns = new_column_names.columns.str.replace("dateCrawled", "date_crawled")
new_column_names.columns = new_column_names.columns.str.replace("postalCode", "post_code")
new_column_names.columns = new_column_names.columns.str.replace("nrOfPictures", "num_pictures")
new_column_names.columns = new_column_names.columns.str.lower()
new_column_names.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'powerps', 'model',
       'odometer', 'registration_month', 'fueltype', 'brand',
       'unrepaired_damage', 'ad_created', 'num_pictures', 'post_code',
       'last_seen'],
      dtype='object')

In [7]:
autos.columns = new_column_names.columns
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  powerps   model   odometer  registration_month fueltype  \
0    manuell      158  andere  150,000km                   3      lpg   
1  automatik      286     7er  150,000km                   6   benzin   
2    manuell      102    golf   70,000km                   7   benzin   
3  automatik       71  fortwo   70,000km                   6   benzin   
4    manuell        0   focus  150,000km                   7   benzin   

        brand unrepaired_damage           ad_created  num_pictures  post_code  \
0     peugeot              nein  2016-03-26 00:00:00             0      79588   
1         bmw              nein  2016-04-04 00:00:00             0      71034   
2  volkswagen              nein  2016-03-26 00:00:00             0      35394   
3       smart              nein  2016-03-12 00:00:00             0      33729   
4        ford              nein  2016-04-01 00:00:00             0      39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

## 4. Cleaning up individual columns
Looking at the data more closely.        
Results from looking at the data:         
The seller column appears to have the same entry 49,999 of 50,000 says "private". Perhaps drop this column?         
Registration Year could be cleaned up. Strange entries.        
Odometer column needs text removed and made to an integer.       

In [8]:
autos.describe(include="all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-23 18:39:34  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox       powerps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fueltype       brand unrepaired_damage  \
count       50000        50000.000000    45518       50000             40171   
unique         13                 NaN        7          40                 2   
top     150,000km                 NaN   benzin  volkswagen              nein   
freq        32424                 NaN    30107       10687             35232   
mean          NaN            5.723360      NaN         NaN               NaN   
std           NaN            3.711984      NaN         NaN               NaN   
min           NaN            0.000000      NaN         NaN               NaN   
25%           NaN            3.000000      NaN         NaN               NaN   
50%           NaN            6.000000      NaN         NaN               NaN   
75%           NaN            9.000000      NaN         NaN               NaN   
max           NaN           12.000000      NaN         NaN               NaN   

                 ad_created  num_pictures     post_code            last_seen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

To cleanup the Price column

In [9]:
autos['price'] = (autos['price']
                            .str.replace("$", "")
                            .str.replace(",", "")
                            .astype(int))


<ipython-input-9-1d4e92c71fe4>:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will*not* be treated as literal strings when regex=True.
  autos['price'] = (autos['price']


To cleanup the odometer column

In [10]:
autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)

autos['odometer_km'] = (autos['odometer_km']
                            .str.replace("km", "")
                            .str.replace(",", "")
                            .astype(int))
autos

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offer_type  price   abtest vehicle_type  registration_year  \
0      privat    Angebot   5000  control          bus               2004   
1      privat    Angebot   8500  control    limousine               1997   
2      privat    Angebot   8990     test    limousine               2009   
3      privat    Angebot   4350  control   kleinwagen               2007   
4      privat    Angebot   1350     test        kombi               2003   
...       ...        ...    ...      ...          ...                ...   
49995  privat    Angebot  24900  control    limousine               2011   
49996  privat    Angebot   1980  control       cabrio               1996   
49997  privat    Angebot  13200     test       cabrio               2014   
49998  privat    Angebot  22900  control        kombi               2013   
49999  privat    Angebot   1250  control    limousine               1996   

         gearbox  powerps   model  odometer_km  registration_month fueltype  \
0        manuell      158  andere       150000                   3      lpg   
1      automatik      286     7er       150000                   6   benzin   
2        manuell      102    golf        70000                   7   benzin   
3      automatik       71  fortwo        70000                   6   benzin   
4        manuell        0   focus       150000                   7   benzin   
...          ...      ...     ...          ...                 ...      ...   
49995  automatik      239      q5       100000                   1   diesel   
49996    manuell       75   astra       150000                   5   benzin   
49997  automatik       69     500         5000                  11   benzin   
49998    manuell      150      a3        40000                  11   diesel   
49999    manuell      101  vectra       150000                   1   benzin   

            brand unrepaired_damage           ad_created  num_pictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       post_code            last_seen  
0          79588  2016-04-06 06:45:54  
1          71034  2016-04-06 14:45:08  
2          35394  2016-04-06 20:15:37  
3          33729  2016-03-15 03:16:28  
4          39218  2016-04-01 14:38:50  
...          ...                  ...  
49995     

## 5. Examining the odometer_km and price columns in detail
Examining the odometer_km and price columns more closely for outliners.       

In [11]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        50000 non-null  object
 1   name                50000 non-null  object
 2   seller              50000 non-null  object
 3   offer_type          50000 non-null  object
 4   price               50000 non-null  int32 
 5   abtest              50000 non-null  object
 6   vehicle_type        44905 non-null  object
 7   registration_year   50000 non-null  int64 
 8   gearbox             47320 non-null  object
 9   powerps             50000 non-null  int64 
 10  model               47242 non-null  object
 11  odometer_km         50000 non-null  int32 
 12  registration_month  50000 non-null  int64 
 13  fueltype            45518 non-null  object
 14  brand               50000 non-null  object
 15  unrepaired_damage   40171 non-null  object
 16  ad_created          50

### Examining the Price column

In [12]:
autos["price"].unique().shape

(2357,)

There are 2357 unique entries for the prices and 13 for odometer

In [13]:
autos["price"].describe().round()

count       50000.0
mean         9840.0
std        481104.0
min             0.0
25%          1100.0
50%          2950.0
75%          7200.0
max      99999999.0
Name: price, dtype: float64

In [14]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price   abtest vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   

     gearbox  powerps   model  odometer_km  registration_month fueltype  \
0    manuell      158  andere       150000                   3      lpg   
1  automatik      286     7er       150000                   6   benzin   
2    manuell      102    golf        70000                   7   benzin   
3  automatik       71  fortwo        70000                   6   benzin   
4    manuell        0   focus       150000                   7   benzin   

        brand unrepaired_damage           ad_created  num_pictures  post_code  \
0     peugeot              nein  2016-03-26 00:00:00             0      79588   
1         bmw              nein  2016-04-04 00:00:00             0      71034   
2  volkswagen              nein  2016-03-26 00:00:00             0      35394   
3       smart              nein  2016-03-12 00:00:00             0      33729   
4        ford              nein  2016-04-01 00:00:00             0      39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

In [15]:
autos['price'].value_counts().sort_index(ascending = False)

99999999       1
27322222       1
12345678       3
11111111       2
10000000       1
            ... 
5              2
3              1
2              3
1            156
0           1421
Name: price, Length: 2357, dtype: int64

#### Lets make a separate list of all the cars above $100,000 so that we can catch the index number of overpriced and unrealistic car prices

In [16]:
Top_priced = autos[(autos["price"] >= 100000) ].sort_index(ascending = False)

In [17]:
Top_priced = Top_priced[["price","name","vehicle_type","model"]]

In [18]:
Top_priced

price                                               name  \
49815    130000  SUCHE_TIPPS___Ford_Mustang_Shelby_GT_350_500_K...   
49668    145000         Porsche_911__993__Carrera_S_"SONDERMODELL"   
49391    109999  Lamborghini_Gallardo_LP560_4_E_Gear_"Callisto_...   
47634   3890000                                        Ferrari_FXX   
47598  12345678  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...   
47337    259000                                    BMW_Z8_roadster   
45387    163500  Porsche_911_GT3__Clubsp._Lift_nur_bis_25.03._v...   
44406    120000  Mercedes_Benz_170_S_Cabriolet_A__Startklar_fue...   
43668    197000  Porsche_993/911_Turbo_WLS_II_Exclusive_S_deuts...   
43282    119900               Porsche_911_Carrera_4S_Cabriolet_PDK   
43049    999999                                      2_VW_Busse_T3   
42221  27322222                                  Leasinguebernahme   
40918    198000                             Porsche_911_991_GT3_RS   
39705  99999999                        Tausch_gegen_gleichwertiges   
39377  12345678                        Tausche_volvo_v40_gegen_van   
38814    119500  Porsche_Porsche_964_TURBO_S_X33_WLS_ab_Werk_35...   
38299    250000                                Glas_BMW_mit_Wasser   
37840    220000                                        Porsche_997   
37585    999990                                Volkswagen_Jetta_GT   
36818    350000                                        Porsche_991   
35923    295000  Porsche_911_Targa_Exclusive_Edition__1_von_15_...   
34723    299000  Porsche_Porsche_911/930_Turbo_3.0__deutsche_Au...   
33884    116000               Porsche_911_Carrera_4S_Cabriolet_PDK   
33638    151990                                 Porsche_911_GT3_RS   
32840    169999  Mercedes_Benz_G_500_7G_TRONIC_Cabrio_Desingno_...   
32185    139997  Porsche_997_GT3_MKII_EIN_SAMMLERSTÜCK_IN_VOLLA...   
29286    104900  Porsche_911_Turbo_PDK__Sport_Chrono_Garantie_D...   
28090    194000  Tesla_Model_X_P90D_Signature_Sondermodel__Neuw...   
27371  12345678                                         Fiat_Punto   
24384  11111111                            Schlachte_Golf_3_gt_tdi   
22947   1234566                         Bmw_530d_zum_ausschlachten   
22673    163991  Porsche_911_991_GT3_LED_Licht_Chrono_Navi_Club...   
22060    114400  Tesla_Model_X90D_Autopilot_Leder_AHK_Kaltwette...   
21783    115991               Porsche_911_Carrera_4S_Cabriolet_PDK   
20351    190000                                    BMW_Z8_roadster   
18509    169000                                    Porsche_991_GT3   
17540    105000   Porsche_930_Turbo_classic_analytics_Gutachten_2+   
17140    180000  Mercedes_Benz_SLS_AMG_Roadster_AMG_SPEEDSHIFT_DCT   
16964    105000                    Bentley_Continental_Supersports   
14715    345000                 Rolls_Royce_Phantom_Drophead_Coupe   
14268    135000   Porsche_997_GT3_Approved_Lift_Unfallfrei_Mj.2011   
12682    265000  Porsche_GT3_RS__PCCB__Lift___grosser_Exklusiv_...   
11433    175000  Audi_R8_V10_plus_5.2_FSI_quattro_S_tronic_Akra...   
11137  10000000  suche_maserati_3200_gt_Zustand_unwichtig_laufe...   
10500    155000                                        Porsche_991   
8232     128000               Porsche_993_S_Schalter_BRD_neuwertig   
7814    1300000                                        Ferrari_F40   
7402     115000  Porsche_911_Carrera_4S_Cabrio_PDK__BOSE__NEU__...   
2897   11111111   Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000   
2751     120000                               Porsche_911___993_4S   
2454     137999                                    Porsche_911_GT3   
1878     129000                                  Porsche_911_Turbo   
514      999999                   Ford_Focus_Turnier_1.6_16V_Style   

      vehicle_type        model  
49815        coupe      mustang  
49668        coupe          911  
49391        coupe          NaN  
47634        coupe          NaN  
47598    limousine       vectra  
47337       cabrio      z

#### Lets delete entries with values of 12345678, 99999999, 11111111, 999999 by first making a list


In [19]:
bad_entries_prices = [12345678, 99999999, 11111111, 999999]

Lets check the Citroens as well

In [20]:
citroens = autos[(autos["brand"] == "citroen")]
citroens = citroens.sort_values(by=["price"], ascending=False)
citroens.head()

date_crawled                                               name  \
42221  2016-03-08 20:39:05                                  Leasinguebernahme   
48944  2016-03-21 15:43:49            Citroën_C2_1.6_16V_Senso_Drive_VTR_Plus   
6476   2016-03-29 18:46:08  Citroen_DS20_Pallas_1973_Halbautomatik_/_wenig...   
21720  2016-03-30 20:58:36       Citroën_C5_Tourer_HDi_200_FAP_Aut._Exclusive   
11315  2016-04-01 17:57:47      Roadster__2CV6__Unikat__Karosserie_kompl._Alu   

       seller offer_type     price   abtest vehicle_type  registration_year  \
42221  privat    Angebot  27322222  control    limousine               2014   
48944  privat    Angebot     33000  control   kleinwagen               2005   
6476   privat    Angebot     25000  control          NaN               1973   
21720  privat    Angebot     22900  control        kombi               2013   
11315  privat    Angebot     21111     test       cabrio               1986   

         gearbox  powerps   model  odometer_km  registration_month fueltype  \
42221    manuell      163      c4        40000                   2   diesel   
48944  automatik      109      c2       100000                   1   benzin   
6476         NaN        0  andere       125000                   0      NaN   
21720  automatik      204      c5        60000                   4   diesel   
11315    manuell       27  andere        10000                   0   benzin   

         brand unrepaired_damage           ad_created  num_pictures  \
42221  citroen               NaN  2016-03-08 00:00:00             0   
48944  citroen              nein  2016-03-21 00:00:00             0   
6476   citroen               NaN  2016-03-29 00:00:00             0   
21720  citroen              nein  2016-03-30 00:00:00             0   
11315  citroen               NaN  2016-04-01 00:00:00             0   

       post_code            last_seen  
42221      76532  2016-03-08 20:39:05  
48944      21077  2016-04-06 13:16:48  
6476       67489  2016-04-06 04:44:34  
21720      46147  2016-04-05 11:46:03  
11315      53175  2016-04-05 13:45:03

In [21]:
sonstige_autos = autos[(autos["brand"] == "sonstige_autos")]
sonstige_autos = sonstige_autos.sort_values(by=["price"], ascending=False)
sonstige_autos.head()

date_crawled                                               name  \
11137  2016-03-29 23:52:57  suche_maserati_3200_gt_Zustand_unwichtig_laufe...   
47634  2016-04-04 21:25:21                                        Ferrari_FXX   
7814   2016-04-04 11:53:31                                        Ferrari_F40   
14715  2016-03-30 08:37:24                 Rolls_Royce_Phantom_Drophead_Coupe   
28090  2016-03-08 19:50:06  Tesla_Model_X_P90D_Signature_Sondermodel__Neuw...   

       seller offer_type     price   abtest vehicle_type  registration_year  \
11137  privat    Angebot  10000000  control        coupe               1960   
47634  privat    Angebot   3890000     test        coupe               2006   
7814   privat    Angebot   1300000  control        coupe               1992   
14715  privat    Angebot    345000  control       cabrio               2012   
28090  privat    Angebot    194000     test          suv               2016   

         gearbox  powerps model  odometer_km  registration_month fueltype  \
11137    manuell      368   NaN       100000                   1   benzin   
47634        NaN      799   NaN         5000                   7      NaN   
7814         NaN        0   NaN        50000                  12      NaN   
14715  automatik      460   NaN        20000                   8   benzin   
28090  automatik        0   NaN         5000                   3  elektro   

                brand unrepaired_damage           ad_created  num_pictures  \
11137  sonstige_autos              nein  2016-03-29 00:00:00             0   
47634  sonstige_autos              nein  2016-04-04 00:00:00             0   
7814   sonstige_autos              nein  2016-04-04 00:00:00             0   
14715  sonstige_autos              nein  2016-03-30 00:00:00             0   
28090  sonstige_autos              nein  2016-03-08 00:00:00             0   

       post_code            last_seen  
11137      73033  2016-04-06 21:18:11  
47634      60313  2016-04-05 12:07:37  
7814       60598  2016-04-05 11:34:11  
14715      73525  2016-04-07 00:16:26  
28090      97990  2016-03-16 04:15:31

#### Making a boolean list of these so that I can extract the 

In [22]:
bad_bool = ((autos["price"] == 12345678) | (autos["price"] == 99999999) | (autos["price"] == 11111111) | (autos["price"] == 999999) | (autos["price"] == 1234566) | (autos["price"] == 27322222) | (autos["price"] == 10000000))   

#### The list of bad entries so far

In [23]:
bad_entries = autos[bad_bool]
bad_entries

date_crawled                                               name  \
514    2016-03-17 09:53:08                   Ford_Focus_Turnier_1.6_16V_Style   
2897   2016-03-12 21:50:57   Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000   
11137  2016-03-29 23:52:57  suche_maserati_3200_gt_Zustand_unwichtig_laufe...   
22947  2016-03-22 12:54:19                         Bmw_530d_zum_ausschlachten   
24384  2016-03-21 13:57:51                            Schlachte_Golf_3_gt_tdi   
27371  2016-03-09 15:45:47                                         Fiat_Punto   
39377  2016-03-08 23:53:51                        Tausche_volvo_v40_gegen_van   
39705  2016-03-22 14:58:27                        Tausch_gegen_gleichwertiges   
42221  2016-03-08 20:39:05                                  Leasinguebernahme   
43049  2016-03-21 19:53:52                                      2_VW_Busse_T3   
47598  2016-03-31 18:56:54  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...   

       seller offer_type     price   abtest vehicle_type  registration_year  \
514    privat    Angebot    999999     test        kombi               2009   
2897   privat    Angebot  11111111     test    limousine               1973   
11137  privat    Angebot  10000000  control        coupe               1960   
22947  privat    Angebot   1234566  control        kombi               1999   
24384  privat    Angebot  11111111     test          NaN               1995   
27371  privat    Angebot  12345678  control          NaN               2017   
39377  privat    Angebot  12345678  control          NaN               2018   
39705  privat    Angebot  99999999  control    limousine               1999   
42221  privat    Angebot  27322222  control    limousine               2014   
43049  privat    Angebot    999999     test          bus               1981   
47598  privat    Angebot  12345678  control    limousine               2001   

         gearbox  powerps        model  odometer_km  registration_month  \
514      manuell      101        focus       125000                   4   
2897     manuell       48       escort        50000                   3   
11137    manuell      368          NaN       100000                   1   
22947  automatik      190          NaN       150000                   2   
24384        NaN        0          NaN       150000                   0   
27371        NaN       95        punto       150000                   0   
39377    manuell       95          v40       150000                   6   
39705  automatik      224     s_klasse       150000                   9   
42221    manuell      163           c4        40000                   2   
43049    manuell       70  transporter       150000                   1   
47598    manuell      101       vectra       150000                   3   

      fueltype           brand unrepaired_damage           ad_created  \
514     benzin            ford              nein  2016-03-17 00:00:00   
2897    benzin            ford              nein  2016-03-12 00:00:00   
11137   benzin  sonstige_autos              nein  2016-03-29 00:00:00   
22947   diesel             bmw               NaN  2016-03-22 00:00:00   
24384      NaN      volkswagen               NaN  2016-03-21 00:00:00   
27371      NaN            fiat               NaN  2016-03-09 00:00:00   
39377      NaN           volvo              nein  2016-03-08 00:00:00   
39705   benzin   mercedes_benz               NaN  2016-03-22 00:00:00   
42221   diesel         citroen               NaN  2016-03-08 00:00:00   
43049   benzin      volkswagen               NaN  2016-03-21 00:00:00   
47598   benzin            opel              nein  2016-03-31 00:00:00   

       num_pictures  post_code            last_seen  
514               0      12205  2016-04-06 07:17:35  
2897              0      94469  2016-03-12 22:45:27  
11137             0      73033  2016-04-06 21:18:11  
22947             0      17454  2016-04-02 03:17:32  
24384             0      18519  2016-03-21 1

#### To remove the bad entries from the autos list

In [24]:
for remove in bad_entries.index:
    autos.drop([remove], inplace = True)

#### The statistics for the price column looks  much better now

In [25]:
autos["price"].describe().round()

count      49989.0
mean        5845.0
std        20876.0
min            0.0
25%         1100.0
50%         2950.0
75%         7200.0
max      3890000.0
Name: price, dtype: float64

#### To look at the lowest values next that are priced under 10 dollars:

In [26]:
cheaply_priced = autos[(autos["price"] <= 10)].sort_index(ascending = False)
cheaply_priced

date_crawled                                               name  \
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49960  2016-03-25 22:51:55            Ford_KA_zu_verschenken_***Reserviert***   
49943  2016-03-16 20:46:08                                         Opel_astra   
49884  2016-03-11 13:55:30               Audi_a6_2.5l__Schnaeppchen_nur_heute   
...                    ...                                                ...   
87     2016-03-29 23:37:22                      Bmw_520_e39_zum_ausschlachten   
80     2016-03-09 15:57:57  Nissan_Primera_Hatchback_1_6_16v_73_Kw___99Ps_...   
71     2016-03-28 19:39:35  Suche_Opel_Astra_F__Corsa_oder_Kadett_E_mit_Re...   
55     2016-03-07 02:47:54                     Mercedes_E320_AMG_zu_Tauschen!   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   

       seller offer_type  price   abtest vehicle_type  registration_year  \
49984  privat    Angebot      0     test          NaN               2000   
49974  privat    Angebot      0  control       cabrio               1983   
49960  privat    Angebot      0  control   kleinwagen               1999   
49943  privat    Angebot      0  control          NaN               2016   
49884  privat    Angebot      0     test        kombi               1999   
...       ...        ...    ...      ...          ...                ...   
87     privat    Angebot      0  control          NaN               2000   
80     privat    Angebot      0  control        coupe               1999   
71     privat    Angebot      0  control          NaN               1990   
55     privat    Angebot      1     test          NaN               2017   
27     privat    Angebot      0  control          NaN               2005   

         gearbox  powerps     model  odometer_km  registration_month fueltype  \
49984        NaN        0       NaN       150000                   0      NaN   
49974    manuell       70      golf       150000                   2   benzin   
49960    manuell       60        ka       150000                   6   benzin   
49943    manuell      101     astra       150000                   8   benzin   
49884    manuell      150        a6       150000                  11   diesel   
...          ...      ...       ...          ...                 ...      ...   
87           NaN        0       5er       150000                   0      NaN   
80       manuell       99   primera       150000                   3   benzin   
71       manuell        0       NaN         5000                   0   benzin   
55     automatik      224  e_klasse       125000                   7   benzin   
27           NaN        0       NaN       150000                   0      NaN   

                brand unrepaired_damage           ad_created  num_pictures  \
49984  sonstige_autos               NaN  2016-03-31 00:00:00             0   
49974      volkswagen              nein  2016-03-20 00:00:00             0   
49960            ford               NaN  2016-03-25 00:00:00             0   
49943            opel               NaN  2016-03-16 00:00:00             0   
49884            audi               NaN  2016-03-11 00:00:00             0   
...               ...               ...                  ...           ...   
87                bmw               NaN  2016-03-29 00:00:00             0   
80             nissan                ja  2016-03-09 00:00:00             0   
71               opel               NaN  2016-03-28 00:00:00             0   
55      mercedes_benz              nein  2016-03-06 00:00:00             0   
27               ford               NaN  2016-03-27 00:00:00             0   

       post_code            last_seen  
49984      12103  2016-04-02 19:44:53  
49974       8209  2016-03-27 19:48:16  
49960      34355  2016-03-25 22:51:55  
49943      89134  2016-03-17 19:44:20  
49884      27711  

#### Lets investigate all of the cars priced at 0
How many of them are there?

In [27]:
tot_num_cars = len(autos)

In [28]:
free_num_cars = len(autos[autos["price"] == 0])

In [29]:
print("There are " + str(free_num_cars) + " cars priced at zero of a total " + str(tot_num_cars) + " cars.")

There are 1421 cars priced at zero of a total 49989 cars.


In [30]:
print(str("{:.2f}".format(free_num_cars / tot_num_cars * 100)) + " % of total car sales are set as free")

2.84 % of total car sales are set as free


Since investigating this, perhaps we should leave them for sellers may want to just have their cars removed from them for free

### Examining the Odometer column

In [31]:
autos['odometer_km'].describe()

count     49989.000000
mean     125733.061273
std       40042.153403
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [32]:
unique_values = autos["odometer_km"].unique()
np.sort(unique_values)

array([  5000,  10000,  20000,  30000,  40000,  50000,  60000,  70000,
        80000,  90000, 100000, 125000, 150000])

All of these values look reasonable

### How many null values are there in each of the subheadings?

In [33]:
autos.isnull().sum()

date_crawled             0
name                     0
seller                   0
offer_type               0
price                    0
abtest                   0
vehicle_type          5092
registration_year        0
gearbox               2678
powerps                  0
model                 2755
odometer_km              0
registration_month       0
fueltype              4479
brand                    0
unrepaired_damage     9823
ad_created               0
num_pictures             0
post_code                0
last_seen                0
dtype: int64

This provides an overview of what columns hold information and might be valuable in assessing car sales. It seems that vehicletype, gearbox, model, fueltype and unrepaired damage to be lacking information, 

## 6. Cleaning up the dates

Here I will concentrate on columns: "date_crawled", "last_seen", "ad_created", "registration_month", "registration_year"                 
There are columns as strings and need to be converted to numerical representation.

In [34]:
autos.iloc[0]

date_crawled                       2016-03-26 17:47:46
name                  Peugeot_807_160_NAVTECH_ON_BOARD
seller                                          privat
offer_type                                     Angebot
price                                             5000
abtest                                         control
vehicle_type                                       bus
registration_year                                 2004
gearbox                                        manuell
powerps                                            158
model                                           andere
odometer_km                                     150000
registration_month                                   3
fueltype                                           lpg
brand                                          peugeot
unrepaired_damage                                 nein
ad_created                         2016-03-26 00:00:00
num_pictures                                         0
post_code 

In [35]:
date_col = ["date_crawled", "last_seen", "ad_created"]
for col in date_col:
    autos[col] = autos[col].str[:10]

The dates have now been shortened so that the times are not present

In [36]:
autos.head()

date_crawled                                               name  seller  \
0   2016-03-26                   Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1   2016-04-04         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2   2016-03-26                         Volkswagen_Golf_1.6_United  privat   
3   2016-03-12  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...  privat   
4   2016-04-01  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...  privat   

  offer_type  price   abtest vehicle_type  registration_year    gearbox  \
0    Angebot   5000  control          bus               2004    manuell   
1    Angebot   8500  control    limousine               1997  automatik   
2    Angebot   8990     test    limousine               2009    manuell   
3    Angebot   4350  control   kleinwagen               2007  automatik   
4    Angebot   1350     test        kombi               2003    manuell   

   powerps   model  odometer_km  registration_month fueltype       brand  \
0      158  andere       150000                   3      lpg     peugeot   
1      286     7er       150000                   6   benzin         bmw   
2      102    golf        70000                   7   benzin  volkswagen   
3       71  fortwo        70000                   6   benzin       smart   
4        0   focus       150000                   7   benzin        ford   

  unrepaired_damage  ad_created  num_pictures  post_code   last_seen  
0              nein  2016-03-26             0      79588  2016-04-06  
1              nein  2016-04-04             0      71034  2016-04-06  
2              nein  2016-03-26             0      35394  2016-04-06  
3              nein  2016-03-12             0      33729  2016-03-15  
4              nein  2016-04-01             0      39218  2016-04-01

In [37]:
for item in date_col:
    autos[item] = pd.to_datetime(autos[item], format="%Y-%m-%d")

Yes. It is now a recognisable numerical value for calculations

In [38]:
autos.loc[0, "date_crawled"]

Timestamp('2016-03-26 00:00:00')

In [39]:
autos.loc[0, "date_crawled"].day_name()

'Saturday'

In [40]:
autos.loc[0, "last_seen"].day_name()

'Wednesday'

To look at just the top 5 rows of the interested headings:

In [41]:
dates = autos[["date_crawled", "last_seen", "ad_created", "registration_month", "registration_year"]]
dates.head()

date_crawled  last_seen ad_created  registration_month  registration_year
0   2016-03-26 2016-04-06 2016-03-26                   3               2004
1   2016-04-04 2016-04-06 2016-04-04                   6               1997
2   2016-03-26 2016-04-06 2016-03-26                   7               2009
3   2016-03-12 2016-03-15 2016-03-12                   6               2007
4   2016-04-01 2016-04-01 2016-04-01                   7               2003

The dtype is now datetime64

In [42]:
dates.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 49989 entries, 0 to 49999
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   date_crawled        49989 non-null  datetime64[ns]
 1   last_seen           49989 non-null  datetime64[ns]
 2   ad_created          49989 non-null  datetime64[ns]
 3   registration_month  49989 non-null  int64         
 4   registration_year   49989 non-null  int64         
dtypes: datetime64[ns](3), int64(2)
memory usage: 3.3 MB


What is the oldest and newest advertisement?

In [43]:
autos["ad_created"].min()

Timestamp('2015-06-11 00:00:00')

In [44]:
autos["ad_created"].max()

Timestamp('2016-04-07 00:00:00')

The oldest entry is from 2015 and the oldest is 2016

To order the entries by newest at the top to oldest at the bottom, use:

In [45]:
autos = autos.sort_values(by=["ad_created"], ascending=False)

Lets look closer at the registration year

In [46]:
autos["registration_year"].describe()

count    49989.000000
mean      2005.075077
std        105.724042
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

There are false values of 9999 to be removed. Lets see these entries.                     
Perhaps a more efficient approach is to select everything that is in the future of the current year which is 2021

In [47]:
bad_bool_list = (autos["registration_year"] >= 2021) | (autos["registration_year"] <= 1801)
bad_list = autos[bad_bool_list]
# print(autos["registration_year"] == 9999)
bad_list

date_crawled                                               name  seller  \
38076   2016-04-04                                 Mercedes_Benz_A180  privat   
49910   2016-04-03  Schoener_fast_neuer_Opel_Mokka_in_Zell_Mosel_m...  privat   
32585   2016-04-02                                         UNFAL_Auto  privat   
10556   2016-04-01                                         UNFAL_Auto  privat   
4549    2016-04-01                                         Kompressor  privat   
27578   2016-03-31                                   VW_GOLF_2_SYNCRO  privat   
22316   2016-03-29              VW_Kaefer.__Zwei_zum_Preis_von_einem.  privat   
27618   2016-03-29  Golf_1_75ps_5911km_Original_Automatik_einer_de...  privat   
4164    2016-03-29  Verkaufe_DESIGN_Streifen_/_Aufkleber_VW__Opel_...  privat   
453     2016-03-28                                         Armee_Jeep  privat   
35238   2016-03-26   Suche_Skoda_Fabia____Skoda_Fabia_Combi_mit_Klima  privat   
33950   2016-03-23  58er_karmann_ghia_lowlight_Kaefer__zum_restaur...  privat   
8012    2016-03-23                      Opel_GT_Karosserie_mit_Brief!  privat   
14341   2016-03-23                                  Hole_kostenlos_ab  privat   
22799   2016-03-20                                  Subaru_Impreza_GT  privat   
13559   2016-03-19          Saab_9000_CSE_Automatik_2_3_ltr._mit_EGSD  privat   
42079   2016-03-17                      APE_50___deutsche_Papiere_!!!  privat   
24511   2016-03-17                        Trabant__wartburg__Ostalgie  privat   
49283   2016-03-15                                         Citroen_HY  privat   
49153   2016-03-12                                        Corsa_c20xe  privat   
6308    2016-03-12                              Kaufe_Autos_jeglicher  privat   
8360    2016-03-11                           Vito_touret_119_Blue_Tec  privat   
25003   2016-03-06              Reo_Vorkriegs_Oldtimer_Rennwagen_1928  privat   
24519   2016-03-05  4x_Winterreifen_auf_Alufelge_der_naechste_Wint...  privat   

      offer_type  price   abtest vehicle_type  registration_year    gearbox  \
38076    Angebot  18000     test          NaN               9999        NaN   
49910    Angebot  22200     test          NaN               9000  automatik   
32585    Angebot    450  control          NaN               1800        NaN   
10556    Angebot    450  control          NaN               1800        NaN   
4549     Angebot   1600     test          NaN               4100        NaN   
27578    Angebot   1700     test          NaN               2800    manuell   
22316    Angebot   1500  control          NaN               1000    manuell   
27618    Angebot   7999     test          NaN               5911  automatik   
4164     Angebot     49  control          NaN               5000        NaN   
453      Angebot   9800     test          NaN               4500    manuell   
35238    Angebot      0  control          NaN               1500        NaN   
33950    Angebot   7999     test          NaN               9999        NaN   
8012     Angebot    700     test          NaN               9999        NaN   
14341    Angebot      0     test          NaN               9999        NaN   
22799    Angebot   9000     test          NaN               5000    manuell   
13559    Angebot      0  control          NaN               9000  automatik   
42079    Angebot    500  control          NaN               4800    manuell   
24511    Angebot    490  control          NaN               1111        NaN   
49283    Angebot   7750  control          NaN               1001        NaN   
49153    Angebot   2500     test          NaN               5000        NaN   
6308     Angebot      0     test          NaN               9996        NaN   
8360     Angebot  42800  control          NaN               6200  automatik   
25003    Angebot   6500  control          NaN               8888        NaN   
24519    Angebot    250     test          NaN               5000        NaN   

    

In [48]:
for remove in bad_list.index:
    autos.drop([remove], inplace = True)

Just to check that these values have been removed

In [49]:
autos = autos.sort_values(by=["registration_year"], ascending=False)
autos.head()

date_crawled                                              name  seller  \
5763    2016-03-05            Suche_Auspuff_Anlage_honda_crx_del_sol  privat   
38342   2016-03-07             Suche_Auto_fuer_die_jagd_zum_tauschen  privat   
49185   2016-03-28             Alfa_Romeo_156_Sportwagon_1_8_T_Spark  privat   
49262   2016-03-26                                       Opel_vectra  privat   
8882    2016-03-15  Volvo_C30_1.6D_130294km__TÜV_bis_02.2018___Leder  privat   

      offer_type  price   abtest vehicle_type  registration_year  gearbox  \
5763     Angebot    100     test          NaN               2019  manuell   
38342    Angebot      0  control          NaN               2019      NaN   
49185    Angebot   1200  control          NaN               2019  manuell   
49262    Angebot   2000     test          NaN               2018      NaN   
8882     Angebot   5300  control          NaN               2018  manuell   

       powerps     model  odometer_km  registration_month fueltype  \
5763       125  cr_reihe        20000                   4      NaN   
38342        0       NaN       150000                   0      NaN   
49185      140       156       150000                   5   benzin   
49262        0       NaN       150000                   0      NaN   
8882       109   c_reihe       150000                   6   diesel   

                brand unrepaired_damage ad_created  num_pictures  post_code  \
5763            honda              nein 2016-03-05             0      18106   
38342  sonstige_autos               NaN 2016-03-07             0      18510   
49185      alfa_romeo              nein 2016-03-28             0      91058   
49262            opel               NaN 2016-03-26             0      65549   
8882            volvo              nein 2016-03-15             0      16949   

       last_seen  
5763  2016-03-06  
38342 2016-04-06  
49185 2016-04-06  
49262 2016-04-06  
8882  2016-03-29

Now the following statistics make more sense

In [50]:
autos["registration_year"].describe()

count    49965.000000
mean      2003.368258
std          7.689636
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2019.000000
Name: registration_year, dtype: float64

## 7. Aggregating data to view the brand column

In [51]:
autos["brand"].unique()

array(['honda', 'sonstige_autos', 'alfa_romeo', 'opel', 'volvo',
       'mercedes_benz', 'renault', 'ford', 'suzuki', 'volkswagen',
       'peugeot', 'seat', 'audi', 'fiat', 'bmw', 'mazda', 'mitsubishi',
       'toyota', 'lada', 'citroen', 'hyundai', 'smart', 'daihatsu',
       'lancia', 'dacia', 'kia', 'nissan', 'skoda', 'daewoo', 'chevrolet',
       'subaru', 'saab', 'jaguar', 'porsche', 'mini', 'land_rover',
       'trabant', 'chrysler', 'rover', 'jeep'], dtype=object)

In [52]:
len(autos["brand"].unique())

40

There are 40 different brands available as listed above. 
What is the most expensive price for each brand?
What are top 10 most expensive brands?

In [53]:
expensive_price_each_brand = {}

brands = autos["brand"].unique()
for each in brands:
    selected_rows = autos[autos["brand"] == each]
    sorted_rows = selected_rows.sort_values("price", ascending=False)
    first_row = sorted_rows.iloc[0]
    price = first_row["price"]
    expensive_price_each_brand[each] = price
    

In [54]:
expensive_price_each_brand

{'honda': 41900,
 'sonstige_autos': 3890000,
 'alfa_romeo': 55000,
 'opel': 38990,
 'volvo': 29500,
 'mercedes_benz': 180000,
 'renault': 93000,
 'ford': 130000,
 'suzuki': 23200,
 'volkswagen': 999990,
 'peugeot': 25500,
 'seat': 33500,
 'audi': 175000,
 'fiat': 25000,
 'bmw': 259000,
 'mazda': 38900,
 'mitsubishi': 37800,
 'toyota': 49500,
 'lada': 6500,
 'citroen': 33000,
 'hyundai': 29999,
 'smart': 16850,
 'daihatsu': 9900,
 'lancia': 29500,
 'dacia': 16350,
 'kia': 28450,
 'nissan': 34999,
 'skoda': 32700,
 'daewoo': 4200,
 'chevrolet': 56500,
 'subaru': 31800,
 'saab': 18900,
 'jaguar': 58900,
 'porsche': 350000,
 'mini': 34500,
 'land_rover': 99000,
 'trabant': 6480,
 'chrysler': 20000,
 'rover': 23500,
 'jeep': 55800}

Now to put it into a descending order:

In [55]:
sorted_list = dict(sorted(expensive_price_each_brand.items(), key=lambda item: item[1], reverse=True))
sorted_list

{'sonstige_autos': 3890000,
 'volkswagen': 999990,
 'porsche': 350000,
 'bmw': 259000,
 'mercedes_benz': 180000,
 'audi': 175000,
 'ford': 130000,
 'land_rover': 99000,
 'renault': 93000,
 'jaguar': 58900,
 'chevrolet': 56500,
 'jeep': 55800,
 'alfa_romeo': 55000,
 'toyota': 49500,
 'honda': 41900,
 'opel': 38990,
 'mazda': 38900,
 'mitsubishi': 37800,
 'nissan': 34999,
 'mini': 34500,
 'seat': 33500,
 'citroen': 33000,
 'skoda': 32700,
 'subaru': 31800,
 'hyundai': 29999,
 'volvo': 29500,
 'lancia': 29500,
 'kia': 28450,
 'peugeot': 25500,
 'fiat': 25000,
 'rover': 23500,
 'suzuki': 23200,
 'chrysler': 20000,
 'saab': 18900,
 'smart': 16850,
 'dacia': 16350,
 'daihatsu': 9900,
 'lada': 6500,
 'trabant': 6480,
 'daewoo': 4200}

In [56]:
top_10 = list(sorted_list.items())[:10]
top_10

[('sonstige_autos', 3890000),
 ('volkswagen', 999990),
 ('porsche', 350000),
 ('bmw', 259000),
 ('mercedes_benz', 180000),
 ('audi', 175000),
 ('ford', 130000),
 ('land_rover', 99000),
 ('renault', 93000),
 ('jaguar', 58900)]

What is the mean price for each brand?

In [57]:
mean_price = {}

brands = autos["brand"].unique()
for each in brands:
    selected_rows = autos[autos["brand"] == each]
    average_values = selected_rows["price"].mean()
    mean_price[each] = average_values
    
mean_price

{'honda': 3889.8596491228072,
 'sonstige_autos': 20154.637707948244,
 'alfa_romeo': 3943.562310030395,
 'opel': 2842.8246289169874,
 'volvo': 4685.548245614035,
 'mercedes_benz': 8380.637920101459,
 'renault': 2351.301996672213,
 'ford': 3626.5429968363533,
 'suzuki': 4038.952218430034,
 'volkswagen': 5252.550561797752,
 'peugeot': 3010.8688186813188,
 'seat': 4223.654255319149,
 'audi': 8965.560354891431,
 'fiat': 2697.6771231828616,
 'bmw': 8028.474479454579,
 'mazda': 3962.542932628798,
 'mitsubishi': 3328.227722772277,
 'toyota': 5097.941653160454,
 'lada': 2476.9032258064517,
 'citroen': 3680.8440629470674,
 'hyundai': 5316.754098360656,
 'smart': 3482.971469329529,
 'daihatsu': 1552.09375,
 'lancia': 3070.3508771929824,
 'dacia': 5897.736434108527,
 'kia': 5707.3258426966295,
 'nissan': 4588.879310344828,
 'skoda': 6313.076433121019,
 'daewoo': 1038.3544303797469,
 'chevrolet': 6432.929328621908,
 'subaru': 3721.5555555555557,
 'saab': 3183.493670886076,
 'jaguar': 11076.50649350

In [58]:
sorted_mean_list = dict(sorted(expensive_price_each_brand.items(), key=lambda item: item[1], reverse=True))
sorted_mean_list

{'sonstige_autos': 3890000,
 'volkswagen': 999990,
 'porsche': 350000,
 'bmw': 259000,
 'mercedes_benz': 180000,
 'audi': 175000,
 'ford': 130000,
 'land_rover': 99000,
 'renault': 93000,
 'jaguar': 58900,
 'chevrolet': 56500,
 'jeep': 55800,
 'alfa_romeo': 55000,
 'toyota': 49500,
 'honda': 41900,
 'opel': 38990,
 'mazda': 38900,
 'mitsubishi': 37800,
 'nissan': 34999,
 'mini': 34500,
 'seat': 33500,
 'citroen': 33000,
 'skoda': 32700,
 'subaru': 31800,
 'hyundai': 29999,
 'volvo': 29500,
 'lancia': 29500,
 'kia': 28450,
 'peugeot': 25500,
 'fiat': 25000,
 'rover': 23500,
 'suzuki': 23200,
 'chrysler': 20000,
 'saab': 18900,
 'smart': 16850,
 'dacia': 16350,
 'daihatsu': 9900,
 'lada': 6500,
 'trabant': 6480,
 'daewoo': 4200}

What is the average km?

In [59]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'powerps', 'model',
       'odometer_km', 'registration_month', 'fueltype', 'brand',
       'unrepaired_damage', 'ad_created', 'num_pictures', 'post_code',
       'last_seen'],
      dtype='object')

Lets use the column heading "odometer_km" for the avergae and sort by descending order

In [60]:
mean_km = {}

brands = autos["brand"].unique()
for each in brands:
    selected_rows = autos[autos["brand"] == each]
    average_values = selected_rows["odometer_km"].mean()
    mean_km[each] = round(average_values)

mean_km = dict(sorted(mean_km.items(), key=lambda item: item[1], reverse=True))
mean_km

{'saab': 143671,
 'volvo': 138607,
 'rover': 136449,
 'chrysler': 133149,
 'bmw': 132541,
 'alfa_romeo': 131109,
 'mercedes_benz': 130933,
 'audi': 129644,
 'opel': 129361,
 'volkswagen': 129008,
 'renault': 128224,
 'jeep': 127523,
 'peugeot': 127352,
 'mitsubishi': 126894,
 'subaru': 125556,
 'mazda': 125132,
 'ford': 124153,
 'honda': 123709,
 'lancia': 123158,
 'seat': 122186,
 'daewoo': 121709,
 'jaguar': 121299,
 'citroen': 120043,
 'nissan': 118979,
 'land_rover': 118333,
 'fiat': 117012,
 'toyota': 115989,
 'daihatsu': 114844,
 'kia': 112640,
 'skoda': 111083,
 'suzuki': 109334,
 'hyundai': 106783,
 'smart': 100756,
 'chevrolet': 99523,
 'porsche': 97364,
 'mini': 89375,
 'sonstige_autos': 87754,
 'lada': 86774,
 'dacia': 84729,
 'trabant': 60065}

Lets make this a series

In [61]:
bmp_series = pd.Series(sorted_mean_list)

In [62]:
bmp_series

sonstige_autos    3890000
volkswagen         999990
porsche            350000
bmw                259000
mercedes_benz      180000
audi               175000
ford               130000
land_rover          99000
renault             93000
jaguar              58900
chevrolet           56500
jeep                55800
alfa_romeo          55000
toyota              49500
honda               41900
opel                38990
mazda               38900
mitsubishi          37800
nissan              34999
mini                34500
seat                33500
citroen             33000
skoda               32700
subaru              31800
hyundai             29999
volvo               29500
lancia              29500
kia                 28450
peugeot             25500
fiat                25000
rover               23500
suzuki              23200
chrysler            20000
saab                18900
smart               16850
dacia               16350
daihatsu             9900
lada                 6500
trabant     

Lets make this a DataFrame

In [63]:
new_df = pd.DataFrame(bmp_series, columns=["mean_price"])
new_df.head()

mean_price
sonstige_autos     3890000
volkswagen          999990
porsche             350000
bmw                 259000
mercedes_benz       180000

Now to add the column of average kms

In [64]:
bmp_series = pd.Series(mean_km)
bmp_series.head()

saab        143671
volvo       138607
rover       136449
chrysler    133149
bmw         132541
dtype: int64

In [65]:
new_df["mean_km"] = bmp_series
new_df.head()

mean_price  mean_km
sonstige_autos     3890000    87754
volkswagen          999990   129008
porsche             350000    97364
bmw                 259000   132541
mercedes_benz       180000   130933

In Summary, this table shows that the mean price and the mean km are not directly proportional even though some think that the less you drive a car, the more its value. Perhaps this is still the case for some calssics though.

## 8. Converting German words to English ones

In [66]:
autos.head()

date_crawled                                              name  seller  \
5763    2016-03-05            Suche_Auspuff_Anlage_honda_crx_del_sol  privat   
38342   2016-03-07             Suche_Auto_fuer_die_jagd_zum_tauschen  privat   
49185   2016-03-28             Alfa_Romeo_156_Sportwagon_1_8_T_Spark  privat   
49262   2016-03-26                                       Opel_vectra  privat   
8882    2016-03-15  Volvo_C30_1.6D_130294km__TÜV_bis_02.2018___Leder  privat   

      offer_type  price   abtest vehicle_type  registration_year  gearbox  \
5763     Angebot    100     test          NaN               2019  manuell   
38342    Angebot      0  control          NaN               2019      NaN   
49185    Angebot   1200  control          NaN               2019  manuell   
49262    Angebot   2000     test          NaN               2018      NaN   
8882     Angebot   5300  control          NaN               2018  manuell   

       powerps     model  odometer_km  registration_month fueltype  \
5763       125  cr_reihe        20000                   4      NaN   
38342        0       NaN       150000                   0      NaN   
49185      140       156       150000                   5   benzin   
49262        0       NaN       150000                   0      NaN   
8882       109   c_reihe       150000                   6   diesel   

                brand unrepaired_damage ad_created  num_pictures  post_code  \
5763            honda              nein 2016-03-05             0      18106   
38342  sonstige_autos               NaN 2016-03-07             0      18510   
49185      alfa_romeo              nein 2016-03-28             0      91058   
49262            opel               NaN 2016-03-26             0      65549   
8882            volvo              nein 2016-03-15             0      16949   

       last_seen  
5763  2016-03-06  
38342 2016-04-06  
49185 2016-04-06  
49262 2016-04-06  
8882  2016-03-29

Lets change "unrepaired_damage" from nein to "no damage"              
Lets also change the np.nan to "no damage" as well
Lets also change the response to the offer type from "Angebot" to "offer"
Also, the seller from "privat" to "private"

In [67]:
autos["unrepaired_damage"] = autos["unrepaired_damage"].replace("nein", "no damage")
autos["unrepaired_damage"] = autos["unrepaired_damage"].replace(np.nan, "no damage")
autos["offer_type"] = autos["offer_type"].replace("Angebot", "offer")
autos["seller"] = autos["seller"].replace("privat", "private")
autos.head()

date_crawled                                              name   seller  \
5763    2016-03-05            Suche_Auspuff_Anlage_honda_crx_del_sol  private   
38342   2016-03-07             Suche_Auto_fuer_die_jagd_zum_tauschen  private   
49185   2016-03-28             Alfa_Romeo_156_Sportwagon_1_8_T_Spark  private   
49262   2016-03-26                                       Opel_vectra  private   
8882    2016-03-15  Volvo_C30_1.6D_130294km__TÜV_bis_02.2018___Leder  private   

      offer_type  price   abtest vehicle_type  registration_year  gearbox  \
5763       offer    100     test          NaN               2019  manuell   
38342      offer      0  control          NaN               2019      NaN   
49185      offer   1200  control          NaN               2019  manuell   
49262      offer   2000     test          NaN               2018      NaN   
8882       offer   5300  control          NaN               2018  manuell   

       powerps     model  odometer_km  registration_month fueltype  \
5763       125  cr_reihe        20000                   4      NaN   
38342        0       NaN       150000                   0      NaN   
49185      140       156       150000                   5   benzin   
49262        0       NaN       150000                   0      NaN   
8882       109   c_reihe       150000                   6   diesel   

                brand unrepaired_damage ad_created  num_pictures  post_code  \
5763            honda         no damage 2016-03-05             0      18106   
38342  sonstige_autos         no damage 2016-03-07             0      18510   
49185      alfa_romeo         no damage 2016-03-28             0      91058   
49262            opel         no damage 2016-03-26             0      65549   
8882            volvo         no damage 2016-03-15             0      16949   

       last_seen  
5763  2016-03-06  
38342 2016-04-06  
49185 2016-04-06  
49262 2016-04-06  
8882  2016-03-29

## Thank you for reading!!